### PR TITLE
Add Reduce Scatter communication op.

### PR DESCRIPTION
**Pull request by MayDomine**

### CODE REVIEW

File: bmtrain/distributed/ops.py [hunk 1]
```diff
@@ -98,6 +104,69 @@ def all_gather(x : torch.Tensor, comm = None):
...
+            input.storage(),
+            output.storage(),
+            op,
+            comm
+        )
+        ctx.op = op
+        output = output.transpose(dim, 0)
+        if op in ["sum", "avg"]:
+            pass
+        elif op in ["max", "min"]:
```
Comment (Achazwl) [inline bmtrain/distributed/ops.py:134]: input output 的 dim 不一样

File: bmtrain/distributed/ops.py [hunk 2]
```diff
@@ -98,6 +104,69 @@ def all_gather(x : torch.Tensor, comm = None):
...
+            ctx.save_for_backward( input != output )
+        else:
+            ctx.save_for_backward( output / input )
+        return output
+
+    @staticmethod
+    def backward(ctx, grad_output):
+        with torch.no_grad():
+            grad_output = OpAllGather.apply(grad_output, ctx.comm).transpose(1, ctx.dim).flatten(0,1).transpose(0, ctx.dim)
```
Comment (Achazwl) [inline bmtrain/distributed/ops.py:143]: 第一个 transpose 应为 transpose(1, ctx.dim+1)

File: tests/test_nccl_backward.py [hunk 3]
```diff
@@ -11,10 +11,19 @@ def test_main():
     ref = y
     for i in range(bmt.world_size()):
         if i != bmt.rank(): ref *= i+1
-    print(x.grad)
     assert_eq(x.grad, ref)
+def test_reducescatter():
+    x = torch.ones((24,), dtype=torch.half, device="cuda").requires_grad_(True)
```
Comment (Achazwl) [inline tests/test_nccl_backward.py:16]: 没有测多维

File: tests/test_nccl_backward.py [hunk 4]
```diff
@@ -11,10 +11,19 @@ def test_main():
...
+    y = bmt.distributed.reduce_scatter(x, "sum")
```
Comment (Achazwl) [inline tests/test_nccl_backward.py:17]: dim!=0 和其他 op 类型也测一下

File: tests/test_nccl_backward.py [hunk 5]
```diff
@@ -11,10 +11,19 @@ def test_main():
...
+    loss = y.sum()
+    loss.backward()
+    ref = torch.ones((24,), dtype=torch.half, device="cuda")
```
Comment (Achazwl) [inline tests/test_nccl_backward.py:20]: 这种 ref 转不转置都没区别，拍不出问题

File: .github/ISSUE_TEMPLATE/bug_report.yml [hunk 6]
```diff
@@ -0,0 +1,79 @@
+
```
Comment (Achazwl) [inline .github/ISSUE_TEMPLATE/bug_report.yml:1]: 三个无关的 yml 文件

**Review by Achazwl — CHANGES_REQUESTED** [review]

File: bmtrain/distributed/ops.py [hunk 7]
```diff
@@ -98,6 +104,65 @@ def all_gather(x : torch.Tensor, comm = None):
...
+        else:
+            ctx.save_for_backward( output / input[rank * input.shape[0]:(rank + 1) * input.shape[0]] )
+        return output
+
+    @staticmethod
+    def backward(ctx, grad_output):
+        with torch.no_grad():
+            grad_output = OpAllGather.apply(grad_output, ctx.comm).flatten(0,1)
+        if ctx.op in ["max", "min", "prod"]:
+            raise NotImplementedError("max min operation now do not support backward")
```
Comment (Achazwl) [inline bmtrain/distributed/ops.py:142]: max/min/prod ...  in ReduceScatter

**Review by zkh2016 — APPROVED** [review]
